crates/autopilot/src/domain/settlement/trade.rs

```diff
@@ -39,102 +43,491 @@ impl Trade {
+                    .checked_div(price_limits.buy.0)
+                    .ok_or(error::Math::DivisionByZero)?;
+                let sold = self
+                    .executed
+                    .0
+                    .checked_mul(prices.buy)
+                    .ok_or(error::Math::Overflow)?
+                    .checked_div(prices.sell)
+                    .ok_or(error::Math::DivisionByZero)?;
+                limit_sell.checked_sub(sold).ok_or(error::Math::Negative)
+            }
+            order::Side::Sell => {
+                // scale limit buy to support partially fillable orders
+                let limit_buy = self
+                    .executed
+                    .0
+                    .checked_mul(price_limits.buy.0)
+                    .ok_or(error::Math::Overflow)?
+                    .checked_div(price_limits.sell.0)
+                    .ok_or(error::Math::DivisionByZero)?;
+                let bought = self
+                    .executed
+                    .0
+                    .checked_mul(prices.sell)
+                    .ok_or(error::Math::Overflow)?
+                    .checked_ceil_div(&prices.buy)
+                    .ok_or(error::Math::DivisionByZero)?;
+                bought.checked_sub(limit_buy).ok_or(error::Math::Negative)
+            }
+        }
+        .map(|surplus| eth::Asset {
+            token: self.surplus_token(),
+            amount: surplus.into(),
+        })
+    }
+
     /// Surplus based on custom clearing prices returns the surplus after all
     /// fees have been applied.
     ///
     /// Denominated in NATIVE token
     pub fn native_surplus(&self, prices: &auction::Prices) -> Result<eth::Ether, Error> {
-        let surplus = self.surplus()?;
+        let surplus = self.surplus_over_limit_price()?;
         let price = prices
             .get(&surplus.token)
             .ok_or(Error::MissingPrice(surplus.token))?;
 
         Ok(price.in_eth(surplus.amount))
     }
 
-    /// Surplus based on custom clearing prices returns the surplus after all
-    /// fees have been applied.
+    /// Total fee (protocol fee + network fee). Equal to a surplus difference
+    /// before and after applying the fees.
+    ///
+    /// Denominated in NATIVE token
+    pub fn native_fee(&self, prices: &auction::Prices) -> Result<eth::Ether, Error> {
+        let fee = self
+            .surplus_over_limit_price_before_fee()?
+            .amount
+            .checked_sub(&self.surplus_over_limit_price()?.amount)
+            .ok_or(error::Math::Negative)?;
+        // We don't have to convert the fee to the sell token, we can just use the fee
+        // in surplus token. This is done just because it was done like this in
+        // the previous implementation
+        let fee_in_sell_token = match self.side {
+            order::Side::Buy => fee,
+            order::Side::Sell => fee
+                .checked_mul(&self.prices.uniform.buy.into())
```

Comment:
For sell order `fee` is in buy token, so we convert it to sell with `fee * sell_price / buy_price`.

> But in any case since we only want the native price it seems easier to understand to me that we'd do prices.get(self.buy.token).ok_or().in_eth(fee) rather than convert it first to the sell token and then convert the sell token to ETH.

I would agree here. However I wanted to keep this PR pure refactoring and not introduce any differences to the old code. Let's create an issue for this and discuss with solver team if this is safe to change.

Reply:
https://github.com/cowprotocol/services/issues/2706